### ISSUE TITLE
[SURE-7122] Excessive WebSocket activity when watching resources with permission by name

### ISSUE THREAD

**Issue by moio** [rancher/rancher#41809]
Internal reference: [SURE-7122](https://jira.suse.com/browse/SURE-7122)

## TL;DR

When a user lists a resource in a namespace in which it has permission to only see certain elements, excessive WebSocket activity between browser and Rancher is created due to resource version synchronization being lost and retried multiple times per second.

This results in excessive memory usage from the Rancher pod (https://github.com/rancher/rancher/issues/41225) and possibly excessive load on the Kubernetes API Server (https://github.com/rancher/rancher/issues/41663).

## Reproducer setup

- install Rancher 2.7.4 with default options

- edit the Rancher deployment, eg via:

```sh
kubectl edit --namespace cattle-system deployment/rancher
```

Edit the main container section to add the `--debug` command line argument (in `args`) and the `CATTLE_WATCH_TIMEOUT_SECONDS=20` environment variable (in `env`). Both changes make it easier to see symptoms without altering functionality significantly.

End result should look like the following:
```yaml
      - args:
        - --http-listen-port=80
        - --https-listen-port=443
        - --add-local=true
        - --debug
        env:
        - name: CATTLE_WATCH_TIMEOUT_SECONDS
          value: "20"
        - name: CATTLE_NAMESPACE
          value: cattle-system
        - name: CATTLE_PEER_SERVICE
          value: rancher
```

Wait until the deployment settles.

- create a test user with username `testuser`:

Users & Authentication -> Users -> Create
  Username: testuser
  New Password: testusertestusertestuser
  Confirm Password: testusertestusertestuser
  Click on Create

- allow the new user to access the local cluster:

Explore Cluster -> local -> Cluster and Project Members -> Cluster Membership -> Add
  Select Member: testuser
  Cluster Permissions: Member
  Click on Create

- create a namespace with two ConfigMaps in it. Give access to `testuser` only to one of them:

```sh
export USER_ID=`kubectl get users -o json | jq --raw-output '.items[] | select(.username=="testuser")| .metadata.name'`

kubectl apply -f - <<EOF
apiVersion: v1
kind: Namespace
metadata:
  annotations:
  labels:
    kubernetes.io/metadata.name: test
  name: test
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: test1
  namespace: test
data:
  a: "1"
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: test2
  namespace: test
data:
  a: "2"
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: test
  name: cm-reader
rules:
- apiGroups: [""]
  resources: ["configmaps"]
  resourceNames: ["test1"]
  verbs: ["get", "watch", "list"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: cm-reader
  namespace: test
subjects:
- kind: User
  name: $USER_ID
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: Role
  name: cm-reader
  apiGroup: rbac.authorization.k8s.io
EOF
```

- Log into Rancher as `testuser`
- list ConfigMaps:

Clusters -> local -> More Resources -> Core -> ConfigMaps

Note: it is expected that only the `test1` ConfigMap should be visible at this point, but not `test2`

- Wait a few minutes

Note: if the test cluster uses self-signed SSL certificates, make sure a WebKit based browser is used such as Chrome or Safari. [Firefox has a bug that prevents websockets to work correctly and that will not allow symptoms to emerge](https://bugzilla.mozilla.org/show_bug.cgi?id=594502) as of December 2023.

## Reproducer outputs

* Observe logs of the `rancher` pod for the presence of the string "too old":

```sh
kubectl logs -f --namespace=cattle-system deployment/rancher | grep "too old"
```

Expected result: up to one line every 20 seconds is produced

Actual result: many lines produced per second

```
2023/06/09 14:30:58 [DEBUG] event watch error: too old resource version: 16581 (19656)
2023/06/09 14:30:58 [DEBUG] event watch error: too old resource version: 16581 (19656)
2023/06/09 14:30:58 [DEBUG] event watch error: too old resource version: 16581 (19656)
2023/06/09 14:30:58 [DEBUG] event watch error: too old resource version: 16581 (19656)
...
```

Navigating away from the ConfigMap list will not stop the error flow, logging out will.

**Comment by richard-cox** [comment]
I've managed to reproduce this given the instructions above.

TL;DR The spamming is caused by the lack of `resource.error` `too old` message from Rancher, that's not something the UI can tackle.

In terms of how the UI handles watching resources over sockets and the `resource.stop` messages from rancher...

1. rancher -- kube socket closes (roughly every 30 minutes or some other specific scenarios)
2. rancher --> browser `resource.stop` message sent for every subscribed resource
3. browser --> rancher `watch` style message sent for every subscribed resource. These will either succeed as the revision is still valid (rancher --> browser `resource.start` message sent) or fail as it's to old so we carry on with the below steps
4. rancher --> browser `resource.stop` AND `resource.error` messages sent for every failed `watch` with a bad revision. The `resource.error` contains `too old` message.
5. browser makes http request to fetch latest resource for every failed `watch`
6. browser --> rancher `watch` style message sent with the latest valid revision returned via the http request

After reproducing the error it seems like rancher is failing to send the `resource.error` `too old` message at step 4. This means the UI tries to `watch` with the stale revision... which is rejected... and so on ad nauseum. 

![image_720](https://github.com/rancher/rancher/assets/18697775/ddc56b2a-750a-4eb7-9c64-3423b2417063).

We'd expect to see something like (ignore the revision `1` used in the re-watch, testing in a borked system)...

![image](https://github.com/rancher/rancher/assets/18697775/52856ecd-d0d2-408a-9307-335322d229a7)

**Comment by gaktive** [comment]
Need to reproduce this with 2.7.5 out.

**Comment by moio** [comment]
Still relevant

**Comment by moio** [comment]
@richard-cox https://github.com/rancher/steve/pull/141 does produce the results you suggested.

Without patch:
![image](https://github.com/rancher/rancher/assets/250541/1f20ab99-497b-4738-9b2e-45a0f534699b)

With patch:
![image](https://github.com/rancher/rancher/assets/250541/f4eca4c3-3f76-4656-860f-8687de507e17)

**Comment by richard-cox** [comment]
@moio Looks good, the `resource.error` with `error: too old` is there for the `configmap` resource and a new re-sub (`resource.start`) is sent with a different resource version. There is though a `resource.stop` for `configmap` very quickly afterwards, and the ui then resubs with the same `resourceVersion` as the first time. Does that `resource.start` --> `resource.stop` pattern continue further down the request log?

**Comment by moio** [comment]
> There is though a resource.stop for configmap very quickly afterwards

Right, at timestamp 21:45:**19.611**

>  [...] and the ui then resubs with the same resourceVersion as the first time

Right, but on different resource types (`namespaces` then `navlinks` at 21:45:**19.589**) after their respective separate `resource.stop` events which happen predictably every 20s (I assume because `CATTLE_WATCH_TIMEOUT_SECONDS=20`, per [reproducer conditions](https://github.com/rancher/rancher/issues/41809#issue-1750069869)).


The actual next resub of `configmaps` happens at 21:45:**39.640** (about 20s later) and uses the updated `resourceVersion` (180993).

I reproduced this once again and confirm the only WS traffic I see _on `configmaps`_ is about every 20s.

Here already filtered:
![image](https://github.com/rancher/rancher/assets/250541/0b1d9d0c-030a-419c-bbeb-f0e58de7bec2)

I uploaded the [HAR files for further inspection here](https://drive.google.com/drive/folders/1V9XuO9-DjzQIAkAvTrVxT6QpvJaEYAZe?usp=sharing).

To my understanding, this seems all good - please do correct me if you see anything suspicious.

**Comment by richard-cox** [comment]
Sorry, i worded my comment badly.

The original resource.stop/resource.error --> resource.start (with new revision number) process for configmaps looks good, so from 21:45:19.605 ending 21:45:19.647. My questions were around the resource.stop for configmaps at 21:45:39.639 which is immediately follow by a resub with the previously subscribed resource version. That looks like the old behaviour.

In the `patched.har` file it's more prominent. Lots of old behaviour (resource.stop --> resrouce.start with bad revision), one new / fixed behaviour and then a few of the older behaviour.

Would this be something to do with the way the dev/test environment is set up to trigger the issue? We can go over this tomorrow in our sync up

**Comment by moio** [comment]
User reported success with the `v2.7.6-debug-41809-2` image.

**Comment by moio** [comment]
@richard-cox: @MSpencer87 discovered another side of this bug and I feel like needing your help to figure out why it happens.

Assuming the same reproduction instructions above, `too old resource version` errors are now (as of `v2.7.6-debug-41809-2`) handled correctly when navigating to the ConfigMap list page. Clicking on an individual ConfigMap to go to the ConfigMap detail page also works correctly.

However at that point **reloading the detail ConfigMap page** gets into an infinite loop of `too old resource version` errors. AFAICS the error is delivered back to the UI, but for some reason dashboard code insists on subscribing for the same old revision over and over:

![image](https://github.com/rancher/rancher/assets/250541/1f8527e9-4ced-4517-8f8f-7ebd4dffe6b1)

Here is the HAR file and a repro video for you: https://drive.google.com/drive/folders/1-EiXH9x5F9AIsj8YFD23rwUs_LdbsPz6?usp=sharing

Could this be a dashboard problem?

**Comment by richard-cox** [comment]
Investigated, and the fix has revealed a bug in the UI which is reproducible after refreshing on an old resource's detail page (probably edit as well - my instance died and can't confirm).

Normally we fetch a list of resources and watch with the revision from that fetch. If that becomes 'too old', minus a few other bits, we make a fresh fetch to get the latest resources and revision... and then watch with that revision.

However if we fetch a single resource, we watch with it's own revision which is probably very old. If so the `too old` message will prompt the UI to re-fetch the resource to get the latest copy and revision... and then watch with that resources revision. That revision isn't going to be the latest ... so will probably be too old again. Rinse and repeat.

Bonus fun. The request to fetch the single resource is most probably namespaced... but that part doesn't make it to the request we make... so the url is broken anyway.

I'll update this comment with a UI gh issue.

**Comment by moio** [comment]
@MSpencer87 given the last comment from @richard-cox are you OK considering this done and addressing the remaining problem you found in the upcoming follow-up issue?

**Comment by moio** [comment]
@MSpencer87 can you report test results here please?

**Comment by MSpencer87** [comment]
I was able to successfully reproduce on 2.7.4 and verify using v2.7.6-debug-41809-2, the watch errors surrounding `too old` resource versions have been resolved. 

Step to reproduce:
1) Bring up HA Rancher on 2.7.4 on rke1 v1.25.9-rancher2-1
2) Setup Rancher following the reproducer steps above:
   - Edit Rancher Deployment with `--debug` and `CATTLE_WATCH_TIMEOUT_SECONDS=20`
   - Create a new testuser with member permisions on the local cluster
   - create a namespace with two ConfigMaps, allow testuser access to only one
3) Log into testuser, open browser console and navigate to view the config maps 
`Clusters -> local -> More Resources -> Core -> ConfigMaps`
4) Following logs, watch errors are obvserved via:
```
kubectl logs -f --namespace=cattle-system deployment/rancher | grep "too old"
```


Steps for validation:
1) Bring up HA Rancher with rancherImageTag=v2.7.6-debug-41809-2
2) Setup Rancher following the reproducer steps above:
    - Edit Rancher Deployment with `--debug` and `CATTLE_WATCH_TIMEOUT_SECONDS=20`
    - Create a new testuser with member permisions on the local cluster
    - create a namespace with two ConfigMaps, allow testuser access to only one
3) Log into testuser, open browser console and navigate to view the config maps 
`Clusters -> local -> More Resources -> Core -> ConfigMaps`
4) Following logs, no watch errors are obvserved via:
```
kubectl logs -f --namespace=cattle-system deployment/rancher | grep "too old"
```



Please advise: Opening browser console to Network tab, filtering for `subscribe?sockId=` and viewing Messages with filter `configmap` will still produce message spam seen below. This also seems to impact memory usage on the Rancher pod.

![Screenshot 2024-03-07 at 10 50 25 PM](https://github.com/rancher/rancher/assets/6477538/70382617-b283-4aee-8568-83b1f8fbb069)

@moio @richard-cox can we confirm that remaining websocket message spam is expected behavior to be addressed with https://github.com/rancher/dashboard/issues/10540 ? If so, this can be closed.

**Comment by richard-cox** [comment]
@MSpencer87 Correct. The fix reveals a UI bug which has been addressed in below versions. Those issues are open but the changes are available in -head builds 
- 2.9 https://github.com/rancher/dashboard/issues/10540
- 2.8 https://github.com/rancher/dashboard/issues/10566

2.7 is https://github.com/rancher/dashboard/issues/10567 but blocked on https://github.com/rancher/rancher/pull/44656

**Comment by git-ival** [comment]
Retested on 2.8.2 -> 2.8.3 and confirmed the same results as @MSpencer87's comment. Also encountered the same websocket "configmap" traffic every ~20 seconds or so.
